src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java

```diff
@@ -655,25 +654,27 @@ else if (option == ImportOrderOption.INFLOW) {
      * @param isStatic whether the token is static or not.
      * @param previous previous non-static but current is static (above), or
      *                  previous static but current is non-static (under).
-     * @param line the line of the current import.
+     * @param ast the line of the current import.
      */
-    private void doVisitToken(FullIdent ident, boolean isStatic, boolean previous, int line) {
+    private void doVisitToken(FullIdent ident, boolean isStatic, boolean previous, DetailAST ast) {
```

Comment:
https://github.com/checkstyle/checkstyle/blob/ae9edbd4a8645c8cc1a99ad6890dd0006700d1cb/src/main/java/com/puppycrawl/tools/checkstyle/api/FullIdent.java#L44
`FullIdent` holds the AST too.

Is there any difference between the new AST passed in and the one in FullIdent? If so, what is the difference? If not, doesn't it make sense to use that instead?

Reply:
@rnveach can you please elaborate me more on this change,

Reply:
You changed the code to pass in the AST, but the AST is available from the `FullIdent`.

I am asking, are the 2 ASTs the same instance? If they are not, explain what each one points to.

Reply:
@rnveach I did a dry run for the code and found that the ast instance(present in ImportOrdercheck.java) traverses the nodes, whereas private DetailAST detailAst(present in the FullIdent.java) points specifically to the import package's first node name.
example:
import java.util.List;
ast:import>java>util>List(traversal)
detailAst:java
Code(in FullIdent.java):
``````
    private void append(DetailAST ast) {
        elements.add(ast.getText());
        if (detailAst == null) {
            detailAst = ast;
        }
    }
``````
this piece of code initializes detailAst to java(or whatever the first name be, say org/com...).
In terms of reference yes these two do point to the same instance but usability differs(pointing to a node). And I gave a try to your point as well, passing FullIdent.getDetailAst() rather than ast to the log can provide more precision as it throws error for import's line no. as well as the column no.(which was always '1'  due to imports alignment). It throws an error for incorrect expected column number:

``````
package org.checkstyle.suppressionxpathfilter.importorder;

import java.io.*;
import java.util.Set;
import static java.lang.Math.PI; // warn

public class SuppressionXpathRegressionImportOrderFour {
	// code
}
``````
``````
Expected :/home/kaustubh/main repo/checkstyle/src/it/resources/org/checkstyle/suppressionxpathfilter/importorder/SuppressionXpathRegressionImportOrderFour.java:5:1: Wrong order for 'java.lang.Math.PI' import.
Actual   :/home/kaustubh/main repo/checkstyle/src/it/resources/org/checkstyle/suppressionxpathfilter/importorder/SuppressionXpathRegressionImportOrderFour.java:5:15: Wrong order for 'java.lang.Math.PI' import.
 ``````
This happened because the passed value is now 'java' at 5:15: (FulIdent.detailAst) rather than 
 'import' at 5:1: (ast)

Should I change these passed values to ident.getDetailAst() or let them remain as ast?

Reply:
> example:
import java.util.List;
ast:import>java>util>List(traversal)
detailAst:java

To confirm, `detailAst` is from FullIdent and `ast` is AST being passed in method parameter?
If so, will `detailAst,getParent()` always point to `ast` then?

Reply:
Yes, detailAst is from FullIdent and ast is the AST passed in method parameter of visitiToken();
No, detailAst.getParent() isn't pointing to ast.(They are pointing to different nodes)

Reply:
````
IMPORT -> import [1:0]
|--DOT -> . [1:16]
|  |--DOT -> . [1:11]
|  |  |--IDENT -> java [1:7]
|  |  `--IDENT -> util [1:12]
|  `--IDENT -> List [1:17]
`--SEMI -> ; [1:21]
````

You are correct. I thought the tree looked different.

Ok, I am good to leave it as using the custom AST in the memory parameter. It makes more sense to have the violation be the import itself.